### PR TITLE
Kamaleev Kemal B12-513

### PR DESCRIPTION
**Pull request by f9ka9**

### CODE REVIEW

File: task_02/src/transition.cpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include "transition.hpp"
+
+int FindZeroOneBoundary(const std::vector<int>& values) {
+  for (int index = 0; index + 1 < static_cast<int>(values.size()); ++index) {
```
Comment (LostPointer) [inline task_02/src/transition.cpp]: Линейный обход — сложность O(N). Задача требует O(log N).

File: task_04/src/stack.cpp [hunk 2]
```diff
@@ -1,21 +1,27 @@
...
-
-void Stack::Push(int value) { data_.push(value); }
+void Stack::Push(int value) { data_.push_back(value); }
 
 int Stack::Pop() {
-  auto result = data_.top();
-  data_.pop();
+  const int result = data_.back();
+  data_.pop_back();
   return result;
```
Comment (LostPointer) [inline task_04/src/stack.cpp:13]: `data_.back()` на пустом векторе — UB. Добавить: `if (data_.empty()) throw std::underflow_error("Stack is empty");`

File: task_03/src/test.cpp [hunk 3]
```diff
@@ -1,4 +1,9 @@
-
 #include <gtest/gtest.h>
 
-TEST(TopologySort, Simple) { ASSERT_EQ(1, 1); }
+#include "phone_combinations.hpp"
+
+TEST(PhoneCombinations, BuildsLexicographicOrder) {
```
Comment (LostPointer) [inline task_03/src/test.cpp:24]: Только 1 тест. Добавить edge cases: пустой ввод (`""`), одна цифра (`"2"`), цифры с четырьмя буквами (`"7"`, `"9"`).

File: task_07/src/test.cpp [hunk 4]
```diff
@@ -1,6 +1,15 @@
-
 #include <gtest/gtest.h>
 
-TEST(TopologySort, Simple) {
-  ASSERT_EQ(1, 1);  // Stack []
+#include "merge_sort.hpp"
+
+TEST(MergeSort, SortsValues) {
```
Comment (LostPointer) [inline task_07/src/test.cpp:23]: Только 2 теста. Добавить: пустой массив, один элемент, уже отсортированный, обратно отсортированный, отрицательные числа.

**Review by LostPointer — CHANGES_REQUESTED** [review]
**PR #44 — Камалеев Кемаль Б12-513**

Задачи 01–14 реализованы. PR в целом высокого качества, несколько замечаний.

---

### 🔴 Критично

**task_02 — O(N) вместо O(log N).** `FindZeroOneBoundary` использует линейный `for`-цикл. Задача требует алгоритм за O(log N).

**task_04 — `Stack::Pop()` и `MinStack::Pop()` без проверки на пустоту.** `data_.back()` на пустом векторе — UB. Добавить проверку перед вызовом.

---

### 🟠 Средние замечания

**task_03 — только 1 тест.** Нет edge cases: пустой ввод, одна цифра, '7'/'9' (четыре буквы).

**task_07 — только 2 теста.** Нет: пустой массив, один элемент, уже отсортированный, отрицательные числа.

---

### ✅ Хорошо

- task_01: `std::optional` вместо magic value, `long long` для предотвращения переполнения при сумме.
- task_05, 06: корректные O(N) реализации, `const&` параметры, чистые имена.
- task_07: `static MergeSortImpl` с буфером — детали реализации не торчат в публичном API.
- task_08: `KthOrderStatistic` принимает вектор по значению — элегантно изолирует изменения.
- task_09: Treap на `std::unique_ptr` — нетривиальная и чистая реализация без утечек памяти.
- task_10: open addressing с `kDeleted`-слотами, `std::out_of_range` при отсутствии ключа.
- task_13: `SelectLessons` принимает вектор по значению — не модифицирует входные данные.
- Везде PascalCase, `#pragma once`, правильные `const&`.

File: task_13/src/lesson_schedule.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#include "lesson_schedule.hpp"
+
+#include <algorithm>
+
+std::vector<Lesson> SelectLessons(std::vector<Lesson> lessons) {
+  std::sort(lessons.begin(), lessons.end(),
+            [](const Lesson& lhs, const Lesson& rhs) {
+              if (lhs.end == rhs.end) {
```
Comment (LostPointer) [inline task_13/src/lesson_schedule.cpp]: Сравнение `double` с `==` ненадёжно. Использовать `std::abs(lhs.end - rhs.end) < 1e-9`.

File: task_13/src/lesson_schedule.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  double last_end = -1e100;
+  for (const Lesson& lesson : lessons) {
+    if (lesson.start >= last_end) {
+      selected.push_back(lesson);
+      last_end = lesson.end;
+    }
+  }
+
+  std::sort(selected.begin(), selected.end(),
+            [](const Lesson& lhs, const Lesson& rhs) {
```
Comment (LostPointer) [inline task_13/src/lesson_schedule.cpp:31]: Сравнение `double` с `==` ненадёжно. Использовать `std::abs(lhs.start - rhs.start) < 1e-9`.

File: task_13/src/test.cpp [hunk 7]
```diff
@@ -1,4 +1,15 @@
-
 #include <gtest/gtest.h>
 
-TEST(TopologySort, Simple) { ASSERT_EQ(1, 1); }
+#include "lesson_schedule.hpp"
+
+TEST(LessonSchedule, SelectsMaximumNonOverlappingSet) {
```
Comment (LostPointer) [inline task_13/src/test.cpp:16]: Только 1 тест. Добавить: пустой ввод (`{}`), один урок, все уроки перекрываются (выбирается 1), ни один не перекрывается (выбираются все).

File: task_14/CMakeLists.txt [hunk 8]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.10)
+﻿cmake_minimum_required(VERSION 3.10)
```
Comment (LostPointer) [inline task_14/CMakeLists.txt]: UTF-8 BOM в начале файла (`\xEF\xBB\xBF` перед `cmake`). Сохранить файл без BOM.

**Review by LostPointer — CHANGES_REQUESTED** [review]
**PR #44 — повторный обзор**

Три замечания исправлены: `data_.empty()` guard в task_04, тесты в task_03 и task_07 — всё добавлено.

---

### 🟠 Средние замечания

**task_13 — сравнение `double` с `==` в компараторах.**
`lhs.end == rhs.end` и `lhs.start == rhs.start` — ненадёжно для `double`. Хранить и сравнивать время как целое (секунды/минуты) или использовать `std::abs(lhs.end - rhs.end) < 1e-9`.

---

### 🟡 Стиль

**task_13 — только 1 тест.**
Нет проверок: пустой ввод, один урок, все уроки перекрываются, ни один не перекрывается.

**task_14/CMakeLists.txt — BOM в начале файла.**
UTF-8 BOM (`\xEF\xBB\xBF`) перед `cmake_minimum_required` — может вызвать ошибки разбора на некоторых системах сборки. Открыть файл в редакторе и сохранить без BOM (UTF-8 без BOM).

---

### ✅ Хорошо

- task_02: бинарный поиск O(log N), `int` вместо `size_t` для избежания underflow.
- task_03: итеративное построение комбинаций через `static const array`, 6 тестов.
- task_04: `Stack::Pop` и `MinStack::Pop` бросают исключения; `MinStack` корректно использует shadow-вектор минимумов.
- task_07: merge sort с `static` helper и отдельным буфером, 6 тестов.
